README.md
# e2e-helper

## Usecase
When writing end-to-end test suite for servers (e.g, web-servers, socket-servers, queue consumers) - one needs to launch the tested service, gather it's logs, and terminate the service once the suite concluded.
This holds true *especially* in CI builds.

This package comes to facilitate all these tasks wrapping them for you with setup and tear-down hooks.
You get these hooks from a factory to which you provide your settings.

While you can use it with ***any*** test library for node - since mocha is the most prominent - it can take for you with it a further step and registger these hooks for you.
It can do so for BDD, TDD, and moch-ui-exports UIs.

## Badges
 - [![Build Status](https://secure.travis-ci.org/osher/e2e-helper.png?branch=master)](http://travis-ci.org/osher/e2e-helper) Tested on latests node versions of 6,7,8
 - [![Coverage Status](https://coveralls.io/repos/github/osher/e2e-helper/badge.svg?branch=master)](https://coveralls.io/github/osher/e2e-helper?branch=master)

# Features list:
1. manages for you the setup and teardown hooks
2. manages tunneling the stdout and stderr outputs to a log file
3. supports termination signals as:
   - kill (SIGINT / SIGTERM)
   - PCI message
4. allows you to run your tests against an already running server - i.e - skip the server launch/kill and log-file management by providing env variable SUT (acronym of System Under Test)
   When provided - the SUT may be the base URL against which tests should be run.
5. allows running in code-coverage mode via `istanbul` by providing env variable COVER as truthful value
6. wide set of configurable options to customize `title`, `log-path`, `cwd`, `args`, and more.
7. facilitation wrappers for mocha tdd/bdd UI and [mocha-ui-exports][1]

[1]: https://www.npmjs.com/package/mocha-ui-exports


# usage

## barebones work with the hooks

The lower level mechanism is the main exported module - which is a function that accepts an  options and provides the hooks the following way.
(full options list bellow)

The returned value is the setup handler, where the teardown handler is found both as a property of the returned setup handler, and as a property of the main exported module.

e.g.:
```
const setup = require('e2e-helper')({
  svc:   'bin/server',
  readyNotice: 'server is started'
})

//setup --> function(done) { ... }
//setup.teardown -->  function(done) { ... }

declare('my service', () => {
   before(setup);
   after(setup.teardown);

   require('./api-1.test.js');
   require('./api-2.test.js');
   require('./api-3.test.js');
})

```

Mind that the actual api test suites are nested in a single parent suite, to make sure the setup/teardown hooks are called in time.
That's not the only way to do it.
You can use the hooks in any of the suites independently - but unless if you make sure not to - this will override the server log output file.

## The Full options list

 - `svc` - string, mandatory, should be relative path. a path to the script that starts the service.
   if you need to provide an absolute path - you may use  `.cwd` to provide the absolute path
   when options is string - it is uderstood as options.svc, applying defaults to all the rest
 - `cwd` - string, optional - the work directory the process should run in. defaults to current dir
 - `env` - flat object, environment-variables to pass to the spawned service. default: {}
 - `logPath` - string, optional - path to logfile. default: `'./e2e.log'`
 - `timeout` - integer, optional - timeout for server setup, default: `10000`
 - `slow` - integer, optional - slow bar indicator for server setup, default: `10000`
 - `readyNotice` - string, optional - message to expect on service output that
   indicates the service is ready. default: `'listening on port'`
 - `args` - array, optional, argumnets to be concatenated to the running command
 - `term_code` - string, optional, the termination message to send to the child, default: `'SIGINT'`
 - `term_ipc` - optional, any value provided will be used to child.send(term_ipc) before escalating
   to term-code. When not provided - starts with term-code
 - `term_timeout` - optional, number, timeout in miliseconds before escalations( ipc->term->kill)
   default: `3000`
 - `coverSvc` - optional, path to cover-tool node-script which should run `.svc` to gather coverage
   default: `'./node_modules/istanbul/lib/cli'`
 - `coverArgs` - optional, array of CLI arguments to the cover tool
   default: `["cover","--dir","./coverage/e2e-test","--handle-sigint"]`
 - `coverIgnore` - optional, array of glob-pattern strings for files to exclude from coverage

## with mocha

The package exports 3 facilitators for mocha, for it's main UIs.

All 3 mocha facilitators are using the same lower-level mechanism to retrieve the setup and teardown hooks, and register them.

To make sure the setup and teardown are called first and last respectively - all the suites are loaded to the same root tests tree.
For this - the passed `options` should include them as well.

### Additional options supported in mocha facilitators
 - `suites` - array of strings - paths relative to `process.cwd()` (or absolute) of suites to run between setup and teardown.
 - `title` - passed as 1st argument to the `describe` that will hold the setup, teardown, and all the passed suites. defaults to `end-to-end`

## with mocha, bdd ui

Assume few e2e suites which are organized per endpoint / page  (however this is just a proposal needed to make sense of the example - you can organize your tests anyhow you like)

Then, provide an index file `~/test-e2e/index.js`

```
require('e2e-helper').mocha_bdd({
  svc:   'bin/server',
  suites: [
    'test-e2e/index.html.test.js',
    'test-e2e/api.GET_pets.test.js',
    'test-e2e/api.GET_pet_petId.test.js',
    'test-e2e/api.PUT_pet_petId.test.js',
    // ... more suites per endpoint / page / URL
  ]
  // ... and the rest of the options you may want to provide
})
```

which is ran as `npm e2e`, which in turn is configured to run as:

```
  "scripts" : {
    "e2e": "mocha test-ete/index.js",
```

**NOTE:** If you want to test your server in few execution modes - you may for example provide few files in the fassion that `test-e2e/index.js` is portrayed here,
and configure your `npm e2e` to run all these test-roots.


there is an alias for short form:
```
require('e2e-helper').bdd({
  ...
```

## with mocha, tdd ui

Almost exactly like `mocha bdd ui`, only that in stead of
```
require('e2e-helper').mocha_bdd({
  ...
```
use:

```
require('e2e-helper').mocha_tdd({
  ...
```
or in short:
```
require('e2e-helper').tdd({
  ...
```

## with mocha, using the mocha-ui-exports plugin

Almost exactly like `mocha bdd ui`, only that in stead of
```
require('e2e-helper').mocha_bdd({
  ...
```
use:
```
module.exports = require('e2e-helper').mocha_ui_exports({
  ...
```
or in short:
```
module.exports = require('e2e-helper').exports({
  ...
```

# Stablity
I've used this utility for years now, but took too long to just publish it.
There may be edge-cases with cover mode on windows - that's the only part I'm yet to be satisfied with.
I don't expect API changes for fixing any of these issues.

# Contribute
- submit working code
- if you add functionality - add tests :)
- don't worry about the style. I tried to add eslint to help with that, although it's not bullet-proof.
  If I'll really need to - I'll ask you to permit me on your fork, I'll help as best I can with styles or with anything else :)

# Lisence
MIT, and that's it :)


CHANGELOG.md
e2e-helper
==========
## 0.9.4
  - fix CHANGELOG.md

## 0.9.3
  - setup - fix - pass error to `done` when SUT fails to start
  - add in Readme.md the undocumented feature: `options.env`
  - include node v10.x v12.x and v14.x in tests

## 0.9.2
  - fix bug: kill timeouts should be `.unref()`ed

## 0.9.1
  - allow better customization for cover mode
  - add eslint definitions

## 0.9.0 - official release

predecessor:

mocha-e2e
==========

## 0.6.0 - working quite well, but I'm not satisfied with the names...


test/fixture/env.js
console.log('starting')

setTimeout(() => {
  console.log('the env var: ', process.env.THE_VAR)
  console.log('[][][] listening on port 4343')
}, 100)

process.on('SIGTERM', process.exit)


test/fixture/err.js
setTimeout(function() {
  console.error('oups')
}, 100)

setInterval(function() { console.log('blip') }, 5000)

process.on('SIGTERM', process.exit)

lib/index.js
'use strict'

const {assign} = Object
const path     = require('path')
const fs       = require('fs')
const run      = require('child_process').spawn

const defaults =
  { title        : 'end-to-end'
  , cwd          : process.cwd()
  , env          : {}
  , logPath      : './e2e.log'
  , readyNotice  : 'listening on port'
  , console
  , args         : []
  , timeout      : 10000
  , slow         : 5000
  , term_code    : 'SIGINT'
  , term_timeout : 3000
  , coverSvc     : './node_modules/istanbul/lib/cli'
  , coverArgs    : ['cover', '--dir', './coverage/e2e-test', '--handle-sigint']
  , coverIgnore  : []
  }


/**
  @description
  returns a before-all mocha handler that setups the server with the provided
  options.

  while being called, it sets `e2e.teardown` with the matching teardown
  function that shuts down the service gracefully.

  @name module:e2e-util
  @function
  @param {string|object} options - when provided as string - it's understood as `options.svc`
  @param {string} options.svc - system under test - path to the script that
     runs the target server
  @param {string} options.logPath - path to log file
  @param {string} options.readyNotice - output line expected on stdout of
     the started service that indicates that the service is ready
  @param {array} options.args - args to concat to the node command
  @param {callback} callback
 */
module.exports =
  assign(
    e2e
  , { mocha_bdd
    , mocha_tdd
    , mocha_ui_exports
    , bdd               : mocha_bdd
    , tdd               : mocha_tdd
    , exports           : mocha_ui_exports
    }
  )

/* for tests only */
e2e.internal =
  { initCtx
  , setup
  , teardown
  }

function e2e(options) {
    const ctx = initCtx(options)

    e2e.teardown =
      ctx_setup.teardown =
        ctx_teardown

    return ctx_setup

    //TRICKY: these must be 'oldschool' functions - mocha relays on the `this` keyword
    function ctx_setup(done) { setup(ctx, this, done) }
    function ctx_teardown(done) { teardown(ctx, this, done) }
}

function mocha_functions_api(options, {suite, setup, teardown, console = options.console || global.console}) {
    suite(
      options.title || defaults.title
    , () => {
          if (process.env.SUT) {
              console.log('test target: ', process.env.SUT)
          } else {
              setup(e2e(options))
              teardown(e2e.teardown)
          }

          options.suites
            .map( p => path.resolve(p) )
            .forEach(global.require /* istanbul ignore next */|| require)
      }
    )
}

function mocha_bdd(options) {
    mocha_functions_api(options
    , { suite:    global.describe
      , setup:    global.before
      , teardown: global.after
      }
    )
}

function mocha_tdd(options) {
    mocha_functions_api(options
    , { suite:    global.suite
      , setup:    global.suiteSetup
      , teardown: global.suiteTeardown
      }
    )
}

function mocha_ui_exports(options) {
    const console = options.console || global.console
    const loadSuite = global.require /* istanbul ignore next */ || require
    const res = {}
    const suite = {}

    process.env.SUT
      ? console.log('test target: ', process.env.SUT)
      : Object.assign(suite
        , { beforeAll: e2e(options)
          , afterAll:  e2e.teardown
          }
        )

    options.suites.forEach(
      endpointTest =>
        assign(suite, loadSuite(path.resolve(endpointTest)))
    )

    res[ options.title || defaults.title ]  = suite

    return res
}

function initCtx(options) {
    if ('object' != typeof options) options = { svc: options }

    options = assign({}, defaults, options)
    options.args = options.args.concat()
    options.coverArgs = options.coverArgs.concat()

    const stringArray = (arr) => Array.isArray(arr) && !arr.find(s => 'string' != typeof s)
    const targetSvc = path.join(String(options.cwd), String(options.svc))

    const msg =
          !options.svc                            && 'options.svc is not provided'
       || 'string' != typeof options.svc          && 'options.svc must be a string'
       || 'string' != typeof options.cwd          && 'options.cwd must be a string'
       || !fs.existsSync(options.cwd)             && options.cwd + ' is not found on disk'
       || !isFileFound(targetSvc)                 && options.svc + ' is not found on disk'
       || !options.logPath                        && 'options.logPath is expected to be a path'
       || 'string' != typeof options.logPath      && 'options.logPath is expected to be a path'
       || !options.readyNotice                    && 'options.readyNotice must be a string'
       || 'string' != typeof options.readyNotice  && 'options.readyNotice must be a string'
       || !Array.isArray(options.args)            && 'options.args must be an array'
       || 'number' != typeof options.timeout      && 'options.timeout must be a number'
       || 'number' != typeof options.slow         && 'options.slow must be a number'
       || options.env
          && 'object' != typeof options.env       && 'options.env, when provided must be an object'
       || !(  'number' == typeof options.term_timeout
           && 0 < options.term_timeout
           )                                      && 'options.term_timeout, when provided - must be a positive number'
       || !~[ 'SIGTERM', 'SIGINT', 'SIGQUIT'
            , 'SIGKILL', 'SIGHUP'
            ].indexOf(options.term_code)         && 'options.term_code, when provided - must be a valid process signal'

       || 'string' != typeof options.coverSvc     && 'options.svc, when provided - must be path node CLI script, such as istanbul'
       || !stringArray(options.coverArgs)         && 'options.coverArgs, when provided - must be an array of CLI arguments'
       || !stringArray(options.coverIgnore)       && 'options.coverIgnore, when provided - must be an array of glob pattern strings'

    if (msg) throw assign(
      new Error(
        [ "e2e-util is expected to be called with valid options"
        , "valid options should be an object with the following attributes"
        , "(or a string see later)"
        , " - svc - string, mandatory, should be relative path. a path to the script that starts the service."
        , "   if you need to provide an absolute path - you may use .cwd  as the absolute path"
        , "   when options is string - it is uderstood as options.svc, applying defaults to all the rest"
        , " - cwd - string, optional - the work directory the process should run in. defaults to current dir"
        , " - logPath - string, optional - path to logfile. default: './e2e.log'"
        , " - timeout - integer, optional - timeout for server setup, default: " + defaults.timeout
        , " - slow - integer, optional - slow bar indicator for server setup, default: " + defaults.timeout
        , " - readyNotice - string, optional - message to expect on service output that"
        , "   indicates the service is ready. default: " + defaults.readyNotice
        , " - args - array, optional, argumnets to be concatenated to the running command"
        , " - term_code - string, optional, the termination message to send to the child, default: " + defaults.term_code
        , " - term_ipc - optional, any value provided will be used to child.send(term_ipc) before escalating"
        , "   to term-code. When not provided - starts with term-code"
        , " - term_timeout - optional, number, timeout in miliseconds before escalations( ipc->term->kill)"
        , "   default: " + defaults.term_timeout
        , " - coverSvc - optional, path to cover-tool node-script which should run .svc to gather coverage"
        , "   default: " + defaults.coverSvc
        , " - coverArgs - optional, array of CLI arguments to the cover tool"
        , "   default: " + JSON.stringify(defaults.coverArgs)
        , " - coverIgnore - optional, array of glob-pattern strings for files to exclude from coverage"
        , "reason: "
        , "  " + msg
        ].join('\n')
      )
    , { options }
    )

    if (process.env.COVER) {
        const args = [options.coverSvc].concat(options.coverArgs)

        options.coverIgnore.forEach((pattern) => args.push('-x', pattern))

        args.push(
          options.svc.match(/\.js$/)
            ? options.svc
            : options.svc + '.js'
        )

        options.args =
          options.args.length
            ? args.concat(['--'], options.args)
            : args
    } else {
        options.args.unshift(options.svc)
    }

    return options

    function isFileFound(path) {
        return fs.existsSync(path)
            || fs.existsSync(path + '.js')
    }
}

/**
  @param {object} ctx - context
  @param {string} ctx.sut - system under test - path to the script that
     runs the target server
  @param {string} ctx.logPath - path to log file
  @param {string} ctx.readyNotice - output line expected on stdout of
     the started target service that indicates that the service is running
  @param {mocha.Test} test - the test context that implements .timetout(n), .slow(n) ....
  @param {callback} done - callback
  @returns {undefined}
 */
function setup(ctx, test, done) {
    try { fs.unlinkSync( ctx.logPath ) } catch(e) {}

    test.timeout(ctx.timeout)
    test.slow(ctx.slow)

    const log =
      ctx.log =
        fs.createWriteStream(ctx.logPath, { flags: 'a'} )

    log.writable = true

    const writeLog = (...args) => log.writable && log.write.apply(log, args)

    const child =
      ctx.child =
        run(process.execPath, ctx.args
        , { env:   assign({}, process.env, ctx.env)
          , cwd:   ctx.cwd
          , stdio: ['pipe', 'pipe', 'pipe', 'ipc']
          }
        )

    /* istanbul ignore next */
    child.on('error', (err) => {
        writeLog('ERR: ', err)
        done(err)
    })

    child.stderr.on('data', (data) => {
        data = data.toString()

        /* istanbul ignore if */
        writeLog('ERR: ' + data)
        if (~data.indexOf('Error: listen EADDRINUSE')) {
            done(new Error('Server could not start: Address In Use'))
            done = null
        }
    })

    child.stdout.on('data', (data) => {
        data = data.toString()
        writeLog(data)

        if (~data.indexOf(ctx.readyNotice)) {
            ctx.console.log('service started: %s', ctx.args.join(' '))
            done()
            done = null
        }
    })

    child.on('exit', (e) => {
        child.exitted = true
        ctx.console.log('\n\nservice termination ended', e || 'OK')
        ctx.log.writable = false
        ctx.log.end(() => {
            child.emit('--over--')
	    e && done && done(new Error('failed to start. exit code:' + e))
        })
    })
}

/**
 @param {object} ctx - the context
 @param {mocha.Test} test - the Test object
 @param {callback} done  - callback
 @returns {undefined}
 */
function teardown(ctx, test, done) {
    /* istanbul ignore if */
    if (ctx.child.exitted) return done()
    test.slow(ctx.term_timeout * 1.5)
    test.timeout(ctx.term_timeout * 3)

    ctx.child.on('--over--', done)
    ctx.term_ipc
      ? ipc()
      : term()

    function ipc() {
        ctx.child.send(ctx.term_ipc)
        setTimeout(term, ctx.term_timeout).unref()
    }

    function term() {
        ctx.child.kill(ctx.term_code)
        setTimeout(kill, ctx.term_timeout).unref()
    }

    function kill() {
        ctx.child.kill('SIGKILL')
    }
}
